File: app/src/main/java/observer/Publisher.java
package observer;

import java.util.ArrayList;
import java.util.List;

import gb.android.notes2.model.NoteListItem;

public class Publisher {
    private List<DataChangeObserver> dataChangeObservers;
    private List<NoteReadyObserver> noteReadyObservers;

    public Publisher() {
        dataChangeObservers = new ArrayList<>();
        noteReadyObservers = new ArrayList<>();
    }

    public void subscribeDataChange(DataChangeObserver observer) {
        dataChangeObservers.add(observer);
    }

    public void unsubscribeDataChange(DataChangeObserver observer) {
        dataChangeObservers.remove(observer);
    }

    public void subscribeNoteReady(NoteReadyObserver observer) {
        noteReadyObservers.add(observer);
    }

    public void unsubscribeNoteReady(NoteReadyObserver observer) {
        noteReadyObservers.remove(observer);
    }

    public void notifyDataChanged() {
        for (DataChangeObserver observer : dataChangeObservers) {
            observer.update();
        }
    }

    public void notifyNoteReady(NoteListItem noteListItem, String text) {
        for (NoteReadyObserver observer : noteReadyObservers) {
            observer.receiveNote(noteListItem, text);
        }
    }
}

File: app/src/main/java/gb/android/notes2/observer/Publisher.kt
package gb.android.notes2.observer

import gb.android.notes2.observer.DataChangeObserver
import gb.android.notes2.observer.NoteReadyObserver
import gb.android.notes2.model.NoteListItem
import java.util.ArrayList

class Publisher {

    private val dataChangeObservers: MutableList<DataChangeObserver>
    private val noteReadyObservers: MutableList<NoteReadyObserver>

    fun subscribeDataChange(observer: DataChangeObserver) {
        dataChangeObservers.add(observer)
    }

    fun unsubscribeDataChange(observer: DataChangeObserver) {
        dataChangeObservers.remove(observer)
    }

    fun subscribeNoteReady(observer: NoteReadyObserver) {
        noteReadyObservers.add(observer)
    }

    fun unsubscribeNoteReady(observer: NoteReadyObserver) {
        noteReadyObservers.remove(observer)
    }

    fun notifyDataChanged() {
        for (observer in dataChangeObservers) {
            observer.update()
        }
    }

    fun notifyNoteReady(noteListItem: NoteListItem?, text: String?) {
        for (observer in noteReadyObservers) {
            observer.receiveNote(noteListItem, text)
        }
    }

    init {
        dataChangeObservers = ArrayList()
        noteReadyObservers = ArrayList()
    }
}